src/PerfView/PerfViewData.cs

```diff
@@ -3388,6 +3388,436 @@ private void CreateCSVFile(List<AspNetRequest> requests, string csvFileName)
+                switch (traceEvent.EventName)
+                {
+                    case _eventRequestStart:
+                        inboundRequestCount++;
+                        if (incompleteRequests.TryGetValue(ANCHostingRequest.GetIndexingKeyFromEvent(traceEvent), out request))
+                        {
+                            // request exists, update its properties
+                            request.StartTimeRelativeMSec = traceEvent.TimeStampRelativeMSec;
+                            request.HasStart = true;
+                            request.Method = (string)traceEvent.PayloadByName("method");
+                            request.Path = (string)traceEvent.PayloadByName("path");
+                            if (String.IsNullOrEmpty(request.Path))
+                                request.Path = "/";
+
+                            // if the request is complete then move it to the completeRequests List
+                            if(request.IsComplete)
+                            {
+                                incompleteRequests.Remove(request.IndexingKey);
+                                completeRequests.Add(request);
+                            }
+                        }
+                        else
+                        {
+                            // request does not exist yet, create and track it
+                            request = new ANCHostingRequest(traceEvent);
+                            request.Method = (string)traceEvent.PayloadByName("method");
+                            request.Path = (string)traceEvent.PayloadByName("path");
+                            if (String.IsNullOrEmpty(request.Path))
+                                request.Path = "/";
+                            request.StartTimeRelativeMSec = traceEvent.TimeStampRelativeMSec;
+                            request.HasStart = true;
+                            incompleteRequests.Add(request.IndexingKey, request);
+                        }
+
+                        break;
+
+                    case _eventRequestStop:
+                        if (incompleteRequests.TryGetValue(ANCHostingRequest.GetIndexingKeyFromEvent(traceEvent), out request))
+                        {
+                            // request exists, update its properties
+                            request.EndTimeRelativeMSec = traceEvent.TimeStampRelativeMSec;
+                            request.HasStop = true;
+
+                            // if the request is complete then move it to the completeRequests List
+                            if (request.IsComplete)
+                            {
+                                incompleteRequests.Remove(request.IndexingKey);
+                                completeRequests.Add(request);
+                            }
+                        }
+                        else
+                        {
+                            // request does not exist yet, create and track it
```

Comment:
If you see a stop event, but there isn't a matching request in `incompleteRequests`, is there a reason to start tracking it?  I would not expect you to see any more events associated with this request.

Reply:
This could be useful if there's a request that was started, let's say 1 second, before the trace was started. Then, if a stop event is encountered 10 seconds later, we know that particular request was active for at least 10 seconds. This info is surfaced to the user along with the ActivityId so they can go inspect any related events for it, perhaps to see why it took at least 10 seconds to complete. So while there probably won't be any more events in this particular view, allowing the user to see it allows them to decide whether or not it's useful to them, IMO.

Reply:
This makes sense to me.  If that's the case, then I recommend removing the `if(request.IsComplete)` branch under the `_eventRequestStart` case.  In this case, it would mean that you saw the stop event before the start event, which shouldn't happen.  If it does, I don't think we want to match them, as likely this is a trace issue or a bug.  If an ActivityID gets re-used, we'd definitely want this to reset things for that ActivityID .

Reply:
> In this case, it would mean that you saw the stop event before the start event, which shouldn't happen

To confirm, are you're saying it's guaranteed that:

1. the events are written to the trace in the order they were emitted, and
2. the callbacks are guaranteed to happen in the order of events in the trace

For # 1 - what if the events left the source process out-of-order and were thus received out-of-order by the tracing tool (i.e. PerfView, dotnet-trace, etc.)? 

Thanks!

Reply:
Point 2 is definitely true - the process of dispatching the events is single threaded.

Point 1 is generally true, but not guaranteed.  If they do leave the source out of order, then yes, I would expect them to be out of order.  Generally though, when we have events that show up out of order, we generally toss them unless we can be sure that the order doesn't matter.  Otherwise, we end up telling a story with the data that isn't correct.

Reply:
understood - thanks! I will remove the `if(request.IsComplete)` block as mentioned.